core/trino-main/src/main/java/io/trino/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -587,50 +603,275 @@ public Type getFinalType()
+                accumulator.addInput(groupIds, page);
+            });
+        }
+    }
+
+    /**
+     * {@link SpillableFinalOnlyGroupedAccumulator} enables spilling for {@link FinalOnlyGroupedAccumulator}
+     */
+    private static class SpillableFinalOnlyGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(SpillableFinalOnlyGroupedAccumulator.class).instanceSize();
+
+        private final FinalOnlyGroupedAccumulator delegate;
+        private final ImmutableList<Type> sourceTypes;
+        private final ImmutableList<Type> accumulatorInputChannelTypes;
+        private final ImmutableList<Integer> accumulatorInputChannels;
+
+        private ObjectBigArray<AccumulatorInput> rawInputs = new ObjectBigArray<>();
+        private IntBigArray groupIdCount = new IntBigArray();
+        private ObjectBigArray<RowBlockBuilder> blockBuilders;
+        private long rawInputsSizeInBytes;
+        private long blockBuildersSizeInBytes;
+        private long rawInputsLength;
+
+        public SpillableFinalOnlyGroupedAccumulator(List<Type> sourceTypes, List<Integer> accumulatorInputChannels, FinalOnlyGroupedAccumulator delegate)
+        {
+            this.sourceTypes = ImmutableList.copyOf(requireNonNull(sourceTypes, "types is null"));
+            this.accumulatorInputChannels = ImmutableList.copyOf(requireNonNull(accumulatorInputChannels, "accumulatorInputChannels is null"));
+            this.delegate = requireNonNull(delegate, "delegate is null");
+            this.accumulatorInputChannelTypes = ImmutableList.copyOf(accumulatorInputChannels.stream()
+                    .map(sourceTypes::get)
+                    .collect(toImmutableList()));
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return INSTANCE_SIZE +
+                    delegate.getEstimatedSize() +
+                    (rawInputs == null ? 0 : rawInputsSizeInBytes + rawInputs.sizeOf()) +
+                    (groupIdCount == null ? 0 : groupIdCount.sizeOf()) +
+                    (blockBuilders == null ? 0 : blockBuildersSizeInBytes + blockBuilders.sizeOf());
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return delegate.getFinalType();
+        }
+
         @Override
         public Type getIntermediateType()
         {
-            throw new UnsupportedOperationException();
+            return new ArrayType(RowType.anonymous(accumulatorInputChannelTypes));
         }
 
         @Override
         public void addInput(GroupByIdBlock groupIdsBlock, Page page)
         {
-            groupCount = max(groupCount, groupIdsBlock.getGroupCount());
-            // Add group id block
-            pagesIndex.addPage(page.appendColumn(groupIdsBlock));
+            verify(rawInputs != null && blockBuilders == null, "rawInputs is null or blockBuilders is not null");
```

Comment:
nit: to simplify debugging i would split it into two messages

`verify(rawInputs != null, "rawInputs is expected to be not null")` and `verify(blockBuilders == null, "blockBuilders is expected to be null")`